File: python/1_Arrays&Hashing/02_ValidAnagram.py
import re


class Solution:
    s = "anagram"
    t = "nagaram"

    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        countS, countT = {}, {}

        for i in range(len(s)):
            countS[s[i]] = 1 + countS.get(s[i], 0)
            countT[t[i]] = 1 + countT.get(t[i], 0)

        print(countS)
        print(countT)

        for key in countS:
            if countS[key] != countT.get(key, 0):
                return False

        return True


solution = Solution()
print(solution.isAnagram(solution.s, solution.t))

File: javascript/1_Arrays&Hashing/02_ValidAnagram.js
//const s = "anagram"
//const t = "nagaram"

//const s = "rat"
//const t = "car"

const s = "a"
const t = "ab"

/**
 * Hash Map - Frequency Counter
 * Time O(N) | Space O(1)
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
const isAnagram = (s, t) => {
    const firstMap = new Map()
    const secondMap = new Map()

    if (s.length !== t.length) {
        return false
    }

    for (const f of s) {
        (firstMap.has(f)) ? firstMap.set(f, firstMap.get(f) + 1) : firstMap.set(f, 1)
    }

    for (const s of t) {
        (secondMap.has(s)) ? secondMap.set(s, secondMap.get(s) + 1) : secondMap.set(s, 1)
    }

    for (const key of firstMap.keys()) {
        if (firstMap.get(key) !== secondMap.get(key)) {
            return false
        }
    }

    return true
};

console.log(isAnagram(s, t))